chalk-solve/src/recursive/lib.rs

```diff
@@ -0,0 +1,194 @@
+use super::search_graph::DepthFirstNumber;
+use chalk_ir::debug;
+use chalk_ir::interner::Interner;
+use chalk_ir::{Canonical, ConstrainedSubst, Goal, InEnvironment, Substitution, UCanonical};
+use std::fmt;
+
+pub type UCanonicalGoal<I> = UCanonical<InEnvironment<Goal<I>>>;
+
+/// The `minimums` struct is used while solving to track whether we encountered
+/// any cycles in the process.
+#[derive(Copy, Clone, Debug)]
+pub struct Minimums {
+    pub positive: DepthFirstNumber,
+}
+
+impl Minimums {
+    pub fn new() -> Self {
+        Minimums {
+            positive: DepthFirstNumber::MAX,
+        }
+    }
+
+    pub fn update_from(&mut self, minimums: Minimums) {
+        self.positive = ::std::cmp::min(self.positive, minimums.positive);
+    }
+}
+
+/// A (possible) solution for a proposed goal.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Solution<I: Interner> {
```

Comment:
This is mostly just copied from `chalk_solve::solve`. But I didn't really think of a good way to handle this. We can make `SearchGraph` generic over `Solution`, but `Fulfill:solve` needs to know about these. I wish it was easier to split out `solve` from `Fulfill` (similar to how we split out `aggregate` from `chalk-engine`, but the `solve` function needs to know a *lot* about the internals of `Fulfill` that we would have to expose publicly.

Reply:
I feel.. hmm.. a bit confused about this. I guess I think I would expect this type to potentially be in chalk-ir -- is it really *specific* to the recursive solver? Don't we use a similar type for SLG? (I guess as the *output* from aggregation?) 

I think my take is that aggregation probably belongs in with the SLG solver in the end, although it is definitely a distinct "layer".